pkg/registry/core/adapters/ns_registry.go

```diff
@@ -14,90 +14,54 @@
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/streamchannel"
 )
 
 type networkServiceRegistryServer struct {
 	client registry.NetworkServiceRegistryClient
 }
 
 func (n *networkServiceRegistryServer) Register(ctx context.Context, request *registry.NetworkService) (*registry.NetworkService, error) {
-	doneCtx := withCapturedContext(ctx)
-	ns, err := n.client.Register(doneCtx, request)
-	if err != nil {
-		return nil, err
-	}
-	lastCtx := getCapturedContext(doneCtx)
-	if lastCtx == nil {
-		return ns, nil
-	}
-	return next.NetworkServiceRegistryServer(ctx).Register(lastCtx, request)
+	return next.NewNetworkServiceRegistryClient(
+		n.client,
+		&nextNSServer{next.NetworkServiceRegistryServer(ctx)},
+	).Register(ctx, request)
 }
 
-func (n *networkServiceRegistryServer) Find(query *registry.NetworkServiceQuery, s registry.NetworkServiceRegistry_FindServer) error {
-	doneCtx := withCapturedContext(s.Context())
-	client, err := n.client.Find(doneCtx, query)
-	if err != nil {
+func (n *networkServiceRegistryServer) Find(query *registry.NetworkServiceQuery, server registry.NetworkServiceRegistry_FindServer) error {
+	client, err := next.NewNetworkServiceRegistryClient(
+		n.client,
+		&nextNSServer{next.NetworkServiceRegistryServer(server.Context())},
+	).Find(server.Context(), query)
+	if client == nil || err != nil {
 		return err
 	}
-	if client == nil {
-		return nil
-	}
-	for {
-		if err := client.Context().Err(); err != nil {
-			break
-		}
-		msg, err := client.Recv()
-		if err != nil {
-			if errors.Is(err, io.EOF) {
-				break
-			}
-			return err
-		}
-		err = s.Send(msg)
-		if err != nil {
-			return err
-		}
-	}
-	lastCtx := getCapturedContext(doneCtx)
-	if lastCtx != nil {
-		return next.NetworkServiceRegistryServer(s.Context()).Find(query, streamcontext.NetworkServiceRegistryFindServer(lastCtx, s))
-	}
-	return nil
+	return nsFindClientToServer(client, server)
 }
 
 func (n *networkServiceRegistryServer) Unregister(ctx context.Context, request *registry.NetworkService) (*empty.Empty, error) {
-	doneCtx := withCapturedContext(ctx)
-	ns, err := n.client.Unregister(doneCtx, request)
-	if err != nil {
-		return nil, err
-	}
-	lastCtx := getCapturedContext(doneCtx)
-	if lastCtx == nil {
-		return ns, nil
-	}
-	return next.NetworkServiceRegistryServer(ctx).Unregister(lastCtx, request)
+	return next.NewNetworkServiceRegistryClient(
+		n.client,
+		&nextNSServer{next.NetworkServiceRegistryServer(ctx)},
+	).Unregister(ctx, request)
 }
 
 // NetworkServiceClientToServer - returns a registry.NetworkServiceRegistryClient wrapped around the supplied client
 func NetworkServiceClientToServer(client registry.NetworkServiceRegistryClient) registry.NetworkServiceRegistryServer {
-	return &networkServiceRegistryServer{client: next.NewNetworkServiceRegistryClient(client, &contextNSClient{})}
+	return &networkServiceRegistryServer{client}
```

Comment:
Note: good practice in go is using named fields on initialization: 
```suggestion
	return &networkServiceRegistryServer{client: client}
```
It could help for readers to understand what is initializing and also could prevent issues with extending struct.